# Exploratory Data Analysis of Healthcare DataSet -Drug Persistency

In [89]:
import pandas as pd

In [90]:
from pandas import read_excel
features = 'Feature Description'
data= 'Dataset'
file_name = 'Healthcare_dataset.xlsx'



In [91]:
df = read_excel(file_name, sheet_name = data,na_values=['Unknown', 'Other/Unknown'])

In [92]:
df.head(10)

Ptid Persistency_Flag  Gender       Race     Ethnicity   Region Age_Bucket  \
0   P1       Persistent    Male  Caucasian  Not Hispanic     West        >75   
1   P2   Non-Persistent    Male      Asian  Not Hispanic     West      55-65   
2   P3   Non-Persistent  Female        NaN      Hispanic  Midwest      65-75   
3   P4   Non-Persistent  Female  Caucasian  Not Hispanic  Midwest        >75   
4   P5   Non-Persistent  Female  Caucasian  Not Hispanic  Midwest        >75   
5   P6   Non-Persistent  Female  Caucasian  Not Hispanic  Midwest        >75   
6   P7   Non-Persistent  Female  Caucasian  Not Hispanic  Midwest        >75   
7   P8   Non-Persistent  Female  Caucasian  Not Hispanic  Midwest        >75   
8   P9   Non-Persistent  Female  Caucasian  Not Hispanic  Midwest        >75   
9  P10   Non-Persistent  Female  Caucasian  Not Hispanic  Midwest        >75   

         Ntm_Speciality Ntm_Specialist_Flag      Ntm_Speciality_Bucket  ...  \
0  GENERAL PRACTITIONER              Others  OB/GYN/Others/PCP/Unknown  ...   
1  GENERAL PRACTITIONER              Others  OB/GYN/Others/PCP/Unknown  ...   
2  GENERAL PRACTITIONER              Others  OB/GYN/Others/PCP/Unknown  ...   
3  GENERAL PRACTITIONER              Others  OB/GYN/Others/PCP/Unknown  ...   
4  GENERAL PRACTITIONER              Others  OB/GYN/Others/PCP/Unknown  ...   
5  GENERAL PRACTITIONER              Others  OB/GYN/Others/PCP/Unknown  ...   
6  GENERAL PRACTITIONER              Others  OB/GYN/Others/PCP/Unknown  ...   
7  GENERAL PRACTITIONER              Others  OB/GYN/Others/PCP/Unknown  ...   
8  GENERAL PRACTITIONER              Others  OB/GYN/Others/PCP/Unknown  ...   
9  GENERAL PRACTITIONER              Others  OB/GYN/Others/PCP/Unknown  ...   

  Risk_Family_History_Of_Osteoporosis Risk_Low_Calcium_Intake  \
0                                   N                       N   
1                                   N                       N   
2                                   N                       Y   
3                                   N                       N   
4                                   N                       N   
5                                   Y                       N   
6                                   N                       N   
7                                   N                       N   
8                                   Y                       N   
9                                   N                       N   

   Risk_Vitamin_D_Insufficiency Risk_Poor_Health_Frailty  \
0                             N                        N   
1                             N                        N   
2                             N                        N   
3                             N                        N   
4                             N                        N   
5                             N                        N   
6                             N                        N   
7                             N                        N   
8                             N                        N   
9                             N                        N   

  Risk_Excessive_Thinness Risk_Hysterectomy_Oophorectomy  \
0                       N                              N   
1                       N                              N   
2                       N                              N   
3                       N                              N   
4                       N                              N   
5                       N                              N   
6                       N                              N   
7                       N                              N   
8                       N                              N   
9                       N                              N   

  Risk_Estrogen_Deficiency Risk_Immobilization Risk_Recurring_Falls  \
0                        N                   N                    N   
1                        N                   N                    N   


## Reading features description

In [93]:
df1 = read_excel(file_name, sheet_name = features)

In [94]:
df1

Bucket                             Variable  \
0              Unique Row Id                           Patient ID   
1            Target Variable                     Persistency_Flag   
2               Demographics                                  Age   
3                        NaN                                 Race   
4                        NaN                               Region   
5                        NaN                            Ethnicity   
6                        NaN                               Gender   
7                        NaN                        IDN Indicator   
8        Provider Attributes            NTM - Physician Specialty   
9           Clinical Factors                       NTM - T-Score    
10                       NaN                   Change in T Score    
11                       NaN                   NTM - Risk Segment   
12                       NaN               Change in Risk Segment   
13                       NaN          NTM - Multiple Risk Factors   
14                       NaN            NTM - Dexa Scan Frequency   
15                       NaN              NTM - Dexa Scan Recency   
16                       NaN                  Dexa During Therapy   
17                       NaN     NTM - Fragility Fracture Recency   
18                       NaN    Fragility Fracture During Therapy   
19                       NaN         NTM - Glucocorticoid Recency   
20                       NaN  Glucocorticoid Usage During Therapy   
21  Disease/Treatment Factor          NTM - Injectable Experience   
22                       NaN                   NTM - Risk Factors   
23                       NaN                   NTM - Comorbidity    
24                       NaN                   NTM - Concomitancy   
25                       NaN                            Adherence   

                                 Variable Description  
0                           Unique ID of each patient  
1   Flag indicating if a patient was persistent or...  
2             Age of the patient during their therapy  
3          Race of the patient from the patient table  
4        Region of the patient from the patient table  
5     Ethnicity of the patient from the patient table  
6        Gender of the patient from the patient table  
7              Flag indicating patients mapped to IDN  
8     Specialty of the HCP that prescribed the NTM Rx  
9   T Score of the patient at the time of the NTM ...  
10  Change in Tscore before starting with any ther...  
11  Risk Segment of the patient at the time of the...  
12  Change in Risk Segment before starting with an...  
13  Flag indicating if  patient falls under multip...  
14  Number of DEXA scans taken prior to the first ...  
15  Flag indicating the presence of Dexa Scan befo...  
16  Flag indicating if the patient had a Dexa Scan...  
17  Flag indicating if the patient had a recent fr...  
18  Flag indicating if the patient had fragility f...  
19  Flag indicating usage of Glucocorticoids (>=7....  
20  Flag indicating if the patient had a Glucocort...  
21  Flag indicating any injectable drug usage in t...  
22  Risk Factors that the patient is falling into....  
23  Comorbidities are divided into two main catego...  
24  Concomitant drugs recorded prior to starting w...  
25                        Adherence for the therapies

In [95]:
df=df.drop(['Ntm_Speciality_Bucket'],axis=1)

### Trying to find out the null values in the data frame

In [96]:
pd.set_option('display.max_rows', None)#to display all the rows

In [97]:
df.isnull().sum()

Ptid                                                                     0
Persistency_Flag                                                         0
Gender                                                                   0
Race                                                                    97
Ethnicity                                                               91
Region                                                                  60
Age_Bucket                                                               0
Ntm_Speciality                                                         310
Ntm_Specialist_Flag                                                      0
Gluco_Record_Prior_Ntm                                                   0
Gluco_Record_During_Rx                                                   0
Dexa_Freq_During_Rx                                                      0
Dexa_During_Rx                                                           0
Frag_Frac_Prior_Ntm      

## we have unknown values in 8 columns as follows:
* race has 97 unknown value
* ehtnicity 91 unknown values
* region has 60 null values
* Ntm_Speciality 310 unknown values
* Risk_Segment_During_Rx  1497 unknown values
* Tscore_Bucket_During_Rx  1497 unknown values
* Change_T_Score 1497unknown values
* Change_Risk_Segment 2229 unknown values

## To find out the percentage of missing data

In [98]:
df.isnull().sum()/len(df)

Ptid                                                                  0.000000
Persistency_Flag                                                      0.000000
Gender                                                                0.000000
Race                                                                  0.028329
Ethnicity                                                             0.026577
Region                                                                0.017523
Age_Bucket                                                            0.000000
Ntm_Speciality                                                        0.090537
Ntm_Specialist_Flag                                                   0.000000
Gluco_Record_Prior_Ntm                                                0.000000
Gluco_Record_During_Rx                                                0.000000
Dexa_Freq_During_Rx                                                   0.000000
Dexa_During_Rx                                      

### To make a categorical dataframe

In [99]:
dfcat=df.copy()

## Its better to drop the column which have more than 40 % of missing value

In [100]:
dfcat.drop('Risk_Segment_During_Rx',inplace=True,axis=1) # 44% data is missing

In [101]:
dfcat.drop('Tscore_Bucket_During_Rx',inplace=True,axis=1)# 44% data is missing

In [102]:
dfcat.drop('Change_T_Score',inplace=True,axis=1)# 44% data is missing

In [103]:
dfcat.drop('Change_Risk_Segment',inplace=True,axis=1)# 65% data is missing

In [104]:
dfcat.head()

Ptid Persistency_Flag  Gender       Race     Ethnicity   Region Age_Bucket  \
0   P1       Persistent    Male  Caucasian  Not Hispanic     West        >75   
1   P2   Non-Persistent    Male      Asian  Not Hispanic     West      55-65   
2   P3   Non-Persistent  Female        NaN      Hispanic  Midwest      65-75   
3   P4   Non-Persistent  Female  Caucasian  Not Hispanic  Midwest        >75   
4   P5   Non-Persistent  Female  Caucasian  Not Hispanic  Midwest        >75   

         Ntm_Speciality Ntm_Specialist_Flag Gluco_Record_Prior_Ntm  ...  \
0  GENERAL PRACTITIONER              Others                      N  ...   
1  GENERAL PRACTITIONER              Others                      N  ...   
2  GENERAL PRACTITIONER              Others                      N  ...   
3  GENERAL PRACTITIONER              Others                      N  ...   
4  GENERAL PRACTITIONER              Others                      Y  ...   

  Risk_Family_History_Of_Osteoporosis  Risk_Low_Calcium_Intake  \
0                                   N                        N   
1                                   N                        N   
2                                   N                        Y   
3                                   N                        N   
4                                   N                        N   

  Risk_Vitamin_D_Insufficiency Risk_Poor_Health_Frailty  \
0                            N                        N   
1                            N                        N   
2                            N                        N   
3                            N                        N   
4                            N                        N   

  Risk_Excessive_Thinness Risk_Hysterectomy_Oophorectomy  \
0                       N                              N   
1                       N                              N   
2                       N                              N   
3                       N                              N   
4                       N                              N   

  Risk_Estrogen_Deficiency Risk_Immobilization Risk_Recurring_Falls  \
0                        N                   N                    N   
1                        N                   N                    N   
2                        N                   N                    N   
3                        N                   N                    N   
4                        N                   N                    N   

  Count_Of_Risks  
0              0  
1              0  
2              2  
3              1  
4              1  

[5 rows x 64 columns]

## for the rest of the missing values i will replace it with most frequent occurring value

In [105]:
dfcat['Race'].value_counts()

Caucasian           3148
African American      95
Asian                 84
Name: Race, dtype: int64

In [106]:
dfcat['Race'].fillna('Caucasian',inplace=True)

In [107]:
dfcat['Ethnicity'].value_counts()

Not Hispanic    3235
Hispanic          98
Name: Ethnicity, dtype: int64

In [108]:
dfcat['Ethnicity'].fillna('Not Hispanic ',inplace=True)

In [109]:
dfcat['Region'].value_counts()

Midwest      1383
South        1247
West          502
Northeast     232
Name: Region, dtype: int64

In [110]:
dfcat['Region'].fillna('Midwest',inplace=True)

In [111]:
dfcat['Ntm_Speciality'].value_counts()

GENERAL PRACTITIONER                                              1535
RHEUMATOLOGY                                                       604
ENDOCRINOLOGY                                                      458
ONCOLOGY                                                           225
OBSTETRICS AND GYNECOLOGY                                           90
UROLOGY                                                             33
ORTHOPEDIC SURGERY                                                  30
CARDIOLOGY                                                          22
PATHOLOGY                                                           16
HEMATOLOGY & ONCOLOGY                                               14
OTOLARYNGOLOGY                                                      14
PEDIATRICS                                                          13
PHYSICAL MEDICINE AND REHABILITATION                                11
SURGERY AND SURGICAL SPECIALTIES                                     8
PULMON

In [112]:
dfcat['Ntm_Speciality'].fillna('GENERAL PRACTITIONER ',inplace=True)

## Check if there is any null value present in the dataframe

In [113]:
dfcat.isnull().sum()

Ptid                                                                  0
Persistency_Flag                                                      0
Gender                                                                0
Race                                                                  0
Ethnicity                                                             0
Region                                                                0
Age_Bucket                                                            0
Ntm_Speciality                                                        0
Ntm_Specialist_Flag                                                   0
Gluco_Record_Prior_Ntm                                                0
Gluco_Record_During_Rx                                                0
Dexa_Freq_During_Rx                                                   0
Dexa_During_Rx                                                        0
Frag_Frac_Prior_Ntm                                             

### Result : there is no null value present in our data frame 